test/cpp/end2end/client_fork_test.cc

```diff
@@ -0,0 +1,176 @@
+#else  // GRPC_ENABLE_FORK_SUPPORT
+
+#include <signal.h>
+
+#include <gtest/gtest.h>
+
+#include "absl/strings/str_cat.h"
+
+#include <grpc/fork.h>
+#include <grpc/grpc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/time.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+#include <grpcpp/server_context.h>
+
+#include "src/core/lib/gprpp/fork.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/util/test_config.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+using ::grpc::testing::EchoRequest;
+using ::grpc::testing::EchoResponse;
+using ::grpc::testing::EchoTestService;
+
+class ServiceImpl final : public EchoTestService::Service {
+  Status BidiStream(
+      ServerContext* /*context*/,
+      ServerReaderWriter<EchoResponse, EchoRequest>* stream) override {
+    EchoRequest request;
+    EchoResponse response;
+    while (stream->Read(&request)) {
+      gpr_log(GPR_INFO, "recv msg %s", request.message().c_str());
+      response.set_message(request.message());
+      stream->Write(response);
+      gpr_log(GPR_INFO, "wrote msg %s", response.message().c_str());
+    }
+    return Status::OK;
+  }
+};
+
+std::unique_ptr<EchoTestService::Stub> MakeStub(const std::string& addr) {
+  return EchoTestService::NewStub(
+      grpc::CreateChannel(addr, InsecureChannelCredentials()));
+}
+
+TEST(ClientForkTest, ClientCallsBeforeAndAfterForkSucceed) {
+  grpc_core::Fork::Enable(true);
+
+  int port = grpc_pick_unused_port_or_die();
+  std::string addr = absl::StrCat("localhost:", port);
+
+  pid_t server_pid = fork();
+  switch (server_pid) {
+    case -1:  // fork failed
+      GTEST_FAIL() << "failure forking";
+    case 0:  // post-fork child
+    {
+      ServiceImpl impl;
+      grpc::ServerBuilder builder;
+      builder.AddListeningPort(addr, grpc::InsecureServerCredentials());
+      builder.RegisterService(&impl);
+      std::unique_ptr<Server> server(builder.BuildAndStart());
+      server->Wait();
+      return;
+    }
+    default:  // post-fork parent
+      break;
+  }
+
+  // Do a round trip before we fork.
+  // NOTE: without this scope, test running with the epoll1 poller will fail.
+  {
+    std::unique_ptr<EchoTestService::Stub> stub = MakeStub(addr);
+    EchoRequest request;
+    EchoResponse response;
+    ClientContext context;
+    context.set_wait_for_ready(true);
+
+    auto stream = stub->BidiStream(&context);
+
+    request.set_message("Hello");
+    ASSERT_TRUE(stream->Write(request));
+    ASSERT_TRUE(stream->Read(&response));
+    ASSERT_EQ(response.message(), request.message());
+  }
+  // Fork and do round trips in the post-fork parent and child.
+  pid_t child_client_pid = fork();
+  switch (child_client_pid) {
+    case -1:  // fork failed
+      FAIL() << "fork failed";
```

Comment:
Please decide on whether to use `GTEST_FAIL` and `FAIL`. I believe `FAIL` is just an alias.

Reply:
Done.